enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,290 @@
+
+* Enable an admin to modify node networking configuration without rebooting (Day 2)
+* Enable cluster administrators to modify node networking after cluster installation (day 2).
+* Integration with MachineConfig node configuration for consistency, familiarity and roll out functionality.
+* Rebootless configuration without deferring errors to a later reboot
+* Rolling back cluster-breaking configuration mistakes.
+
+### Non-Goals
+
+* Replace SRIOV operator
+
+## Proposal
+
+#### Option A
+
+The node networking configuration shall be driven by MCO (and by extension kubernetes-nmstate) and executed by
+[nmstate](https://nmstate.github.io/).
+
+Do a KubeletConfig like integration:
+
+* NodeNetworkState will be introduced. This allows the user to see the network state of a node in the same declarative language that can be used to configure it.
+* The user will configure with NodeNetworkConfigurationPolicy (NNCP). This is a declarative method of defining network state, to be applied by nmstate.
+* NNCP, just like KubeletConfig, targets a MachinePool
+* The MachineConfigDaemon (**MCD**) vendors the knmstate configuring/rollback functionality.
+* MCD runs a vendored subprocess that keeps NodeNetworkState (**NNS**) up to date with the OpenShift API
+* NMState is invoked with Podman (Python project with dependencies outside of stdlib)
+
+1. User Creates **NNCP**
+2. KNMStateController takes the **NNCP** desiredConfig and puts it as a **MachineConfig** (MC) for the desired pool
+3. MachineConfigController renders a new config and its update subcontroller starts updating desiredState of the Nodes
+4. MachineConfigDaemon sees the diff between currentState and desiredState
+5. In the case all the new rendered MC is equal except the **networkState** section, instead of reboot 
+    * MCD writes networkState as is (nmstate valid yaml) and writes it to /tmp
+    * MCD calls nmstatectl set --no-commit /tmp/uuid-of-the-desiredState.yaml
+    * If no connectivity loss, MCD calls nmstatectl commit [PREVIOUS CHECKPOINT]
+    * Otherwise, MCD calls nmstatectl rollback
+* MCD sets the currentState to match the desiredState.
+
+#### Option B
+
+Interact with kubernetes-nmstate handler through MCD. Prepare hooks in MCD for that. User would create MCs, those would be propagated to MCD, MCD would delegate to kubernetes-nmstate handler.
+
+#### Option C
+
+1. Set the NetworkManager config to point to a "merged" keyfile path.
+2. Create two temporary directories for overlay purposes.
+3. Mount an overlay combining the NetworkManager standard /etc/NetworkManager/system-connections and mount it at the path pointed to by 1. e.g., /etc/NetworkManager/system-connections-merged
+
+kubernetes-nmstate will then operate as is, but the keyfiles that ultimately are written by nmstate would effectively be ephemeral. When a node is rebooted, the kubernetes-nmstate-handler will re-process any existing NodeNetworkConfigurationPolicy CRs, and put the configuration back in place.
```

Comment:
Aside but recently I wrote: https://blog.verbum.org/2020/08/22/immutable-%E2%86%92-reprovisionable-anti-hysteresis/

It's not that we want the configuration written by nmstate to be ephemeral per se - it's that the "source of truth" should live in one place - in this case, etcd (and not `/etc`).  By having the implementation of this be ephemeral we're avoiding hysteresis during a node boot - there's no dependency during the boot on the *previous* configuration.